user-ops-indexer/user-ops-indexer/src/indexer/v06/indexer.rs

```diff
@@ -0,0 +1,411 @@
+        tracing::debug!("fetching latest block number");
+        let block_number = self.client.get_block_number().await?.as_u32();
+        tracing::info!(block_number, "latest block number");
+
+        let missed_db_txs = if past_db_logs_start_block != 0 || past_db_logs_end_block != 0 {
+            let from_block = if past_db_logs_start_block > 0 {
+                past_db_logs_start_block as u64
+            } else {
+                max(block_number as i32 + past_db_logs_start_block, 0) as u64
+            };
+            let to_block = if past_db_logs_end_block > 0 {
+                past_db_logs_end_block as u64
+            } else {
+                max(
+                    (block_number - past_rpc_logs_range) as i32 + past_db_logs_end_block,
+                    0,
+                ) as u64
+            };
+            tracing::info!(from_block, to_block, "fetching missed tx hashes in db");
+            let txs = repository::user_op::find_unprocessed_logs_tx_hashes(
+                &self.db,
+                *ENTRYPOINT_V06,
+                UserOperationEventFilter::signature(),
+                from_block,
+                to_block,
+            )
+            .await?;
+            tracing::info!(count = txs.len(), "found missed txs in db");
+            txs
+        } else {
+            Vec::new()
+        };
+
+        let recent_logs = if past_rpc_logs_range > 0 {
+            let from_block = (block_number + 1 - past_rpc_logs_range).max(0);
+            tracing::info!(
+                from_block,
+                to_block = block_number,
+                "fetching past BeforeExecution logs from rpc"
+            );
+            let filter = filter.from_block(from_block).to_block(block_number);
+            let logs = self.client.get_logs(&filter).await?;
+            tracing::info!(count = logs.len(), "fetched past BeforeExecution logs");
+            logs
+        } else {
+            Vec::new()
+        };
+
+        let stream_logs = stream::iter(recent_logs).chain(stream_logs);
+
+        // map to transactions hashes containing user ops, with deduplicated transaction hashes
+        // e.g. [A, A, B, B, B, C, C] -> [A, B, C]
+        let stream_txs = stream::iter(missed_db_txs)
+            .chain(stream_logs.filter_map(|log| async move { log.transaction_hash }))
+            .scan(H256::zero(), |prev, tx_hash| {
+                if *prev == tx_hash {
+                    future::ready(Some(None))
+                } else {
+                    *prev = tx_hash;
+                    future::ready(Some(Some(tx_hash.clone())))
+                }
+            })
+            .filter_map(|tx_hash| async move { tx_hash });
+
+        stream_txs
+            .for_each(|tx| async move {
+                if let Err(err) = &self.handle_tx(tx.clone()).await {
+                    tracing::error!(error = ?err, tx_hash = ?tx, "tx handler failed, skipping");
+                }
+                ()
+            })
+            .await;
```

Comment:
I also thought about writing some custom batching logic, so that transactions that arrive within e.g. 100ms are processed together and written to the DB in the same batch.

Reply:
Yeap, I believe that may be a good idea. We haven't had such a logic before, but if interested some batching ideas have been implemented in [reth-indexer](https://github.com/joshstevens19/reth-indexer/tree/master)

Reply:
Implemented `for_each_concurrent` approach with a constant number of retries, should be a good-enough solution so far.